### PR TITLE
changed broadcast! into bitarray algorithm

### PR DESCRIPTION
**Pull request by chethega**
Cf https://discourse.julialang.org/t/broadcast-vs-slow-performance-allocations/24259/6 for some more discussion and https://github.com/JuliaLang/julia/issues/32047 for the question of validity in view of exceptions. 

@mbauman is this valid with respect to `eachindex`, etc?

Before:
```
julia> using BenchmarkTools, Random
julia> y=1; xsmall=[1]; Random.seed!(42); xlarge=rand(1:4, 100_003);
julia> @btime broadcast(==, $xsmall, $y); @btime  broadcast(==, $xlarge, $y); @show hash(broadcast(==, xlarge, y).chunks);
  860.500 ns (3 allocations: 4.31 KiB)
  152.971 μs (3 allocations: 16.59 KiB)
hash((broadcast(==, xlarge, y)).chunks) = 0xaa3b5a692968e128
```
After:
```
julia> @btime broadcast(==, $xsmall, $y); @btime  broadcast(==, $xlarge, $y); @show hash(broadcast(==, xlarge, y).chunks);
  65.466 ns (2 allocations: 128 bytes)
  42.927 μs (2 allocations: 12.41 KiB)
hash((broadcast(==, xlarge, y)).chunks) = 0xaa3b5a692968e128
```
Monkeypatch:
```
julia> @eval Base.Broadcast @inline function copyto!(dest::BitArray, bc::Broadcasted{Nothing})
           axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
           ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
           destc = dest.chunks
           ind = cind = 1
           bcp = preprocess(dest, bc)
           length(bcp)<=0 && return dest
           @inbounds for i = 0:Base.num_bit_chunks(length(bcp))-2
               z = UInt64(0)
               for j=0:63
                  z |= (bcp[i*64 + j + 1]::Bool) << (j&63)
               end
               destc[i+1] = z
           end
           i = Base.num_bit_chunks(length(bcp))-1
           z = UInt64(0)
           @inbounds for j=0:(length(bcp)-1)&63
                z |= (bcp[i*64 + j + 1]::Bool) << (j&63)
           end
           @inbounds destc[i+1] = z
           return dest
       end
```

### CODE REVIEW

**Comment by chethega** [comment]
Ok, this does not work for cartesian indices. Multidimensional variants I tried:
```
julia> @eval Base.Broadcast @inline function copyto!(dest::BitArray, bc::Broadcasted{Nothing})
           axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
           ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
           destc = dest.chunks
           bcp = preprocess(dest, bc)
           length(destc)<=0 && return dest
           ea = eachindex(bcp)
           i1 = 1
           i2 = 0
           z = UInt64(0)
           @inbounds for idx in ea # = 0:Base.num_bit_chunks(length(bcp))-2
               z |= (bcp[idx]::Bool) << (i2&63)
               i2 += 1
               if (i2 & 63) == 0
                   destc[i1]=z
                   i2 = 0
                   i1 += 1
                   z = UInt64(0)
               end
           end
           i2 != 0 && @inbounds destc[i1] = z
           return dest
       end
```
and
```
julia> @eval Base.Broadcast @inline function copyto!(dest::BitArray, bc::Broadcasted{Nothing})
           axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
           ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
           destc = dest.chunks
           bcp = preprocess(dest, bc)
           length(bcp)<=0 && return dest
           ea = eachindex(bcp)
           idx = first(ea)
           @inbounds for i = 0:Base.num_bit_chunks(length(bcp))-2
               z = UInt64(0)
               for j=0:63
                  z |= (bcp[idx]::Bool) << (j&63)
                  idx = nextind(ea, idx)
               end
               destc[i+1] = z
           end
           i = Base.num_bit_chunks(length(bcp))-1
           z = UInt64(0)
           @inbounds for j=0:(length(bcp)-1)&63
                z |= (bcp[idx]::Bool) << (j&63)
                idx = nextind(ea, idx)
           end
           @inbounds destc[i+1] = z
           return dest
       end
```
None of these present an unambiguous win on my machine. So I restricted the new alg to bitvectors, for the time being. Any ideas how to get the fast variant for higher-dimensional arrays? 

The best way would be to propagate `IndexLinear` information. The second best way would be to find a way to only cache 64 values in a register, instead of an array of Bools (I think the main perf reason for caching is to avoid memory-carried dependencies). Both of my above variants do this, but fail to consistently outperform the current implementation.

**Comment by mbauman** [comment]
This is great — thanks.  Definitely an improvement over the bitcache.  And I do think it should be possible to get this fast for n-dimensional broadcasts, too, but it is indeed tricky.  This gets us a bit closer:

```
    @inbounds for i in 1:Base.num_bit_chunks(length(bcp))-1
        z = UInt64(0)
        for j in 0:63
            z |= bcp[idx] << j
            (idx, s) = iterate(ea, s)
        end
        destc[i]=z
    end
    z = UInt64(0)
    for j in 0:63
        z |= bcp[idx] << j
        r = iterate(ea, s)
        r === nothing && break
        (idx, s) = r
    end
    destc[end]=z
```

But the trouble is that we're still introducing a branch with `(idx, s) = iterate(ea, s)` since that _could_ return `nothing` and we can't really turn that off. We could potentially just use `nextind` instead of iterate, which I think should give us the performance we want.

**Comment by chethega** [comment]
If we somehow managed to get linear indexing working for common constructions, then I think that we could maybe remove the bitcache logic entirely (also from bitarray constructors), replacing it by caching a single `UInt64` (reducing complexity and linecount, yay!). If linear indexing is genuinely unsupported, then the current slight perf regression for that should not matter (would need benchmarking). And linear indexing would be a giant win for more important broadcast operations than bitarrays, cf. [32051](https://github.com/JuliaLang/julia/issues/32051) for a 20 x speedup of `arr .+ const` between `N` and `1 x N` sizes.
```
julia> @eval Base.Broadcast @inline function copyto!(dest::BitArray, bc::Broadcasted{Nothing})
           axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
           ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
           destc = dest.chunks
           bcp = preprocess(dest, bc)
           length(destc)<=0 && return dest
           ea = eachindex(bcp)
           i1 = 1
           i2 = 0
           z = UInt64(0)
           @inbounds for idx in ea
               z |= (bcp[idx]::Bool) << (i2&63)
               i2 += 1
               if i2 == 64
                   destc[i1]=z
                   i2 = 0
                   i1 += 1
                   z = UInt64(0)
               end
           end
           i2 != 0 && @inbounds destc[i1] = z
           return dest
       end
```

File: base/broadcast.jl [hunk 1]
```diff
@@ -890,6 +890,30 @@ preprocess_args(dest, args::Tuple{}) = ()
...
+@inline function copyto!(dest::BitVector, bc::Broadcasted{Nothing})
+    axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
+    ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
+    destc = dest.chunks
+    length(destc)<=0 && return dest
+    bcp = preprocess(dest, bc)
+    @inbounds for i = 0:length(destc)-2
+        z = UInt64(0)
+        for j=0:63
+           z |= (bcp[i*64 + j + 1]::Bool) << (j&63)
```
Comment (mbauman) [inline base/broadcast.jl]: Why the `& 63`?

Comment (StefanKarpinski) [reply base/broadcast.jl]: It avoids emitting code that checks if the shift value is between 0 and 63.

Comment (StefanKarpinski) [reply base/broadcast.jl]: Oh right, I see what you mean, it should probably be `&5` or something like that.

Comment (mbauman) [reply base/broadcast.jl]: I figured as such — in that case 63 is correct.

```llvm
julia> f(x, y) = x << y
f (generic function with 1 method)

julia> g(x, y) = x << (y & 63)
g (generic function with 1 method)

julia> @code_llvm debuginfo=:none f(UInt(1), 1)

define i64 @julia_f_15983(i64, i64) {
top:
  %2 = icmp slt i64 %1, 0
  %3 = icmp ugt i64 %1, 63
  %4 = shl i64 %0, %1
  %5 = select i1 %3, i64 0, i64 %4
  %6 = sub i64 0, %1
  %7 = icmp ugt i64 %6, 63
  %8 = lshr i64 %0, %6
  %9 = select i1 %7, i64 0, i64 %8
  %10 = select i1 %2, i64 %9, i64 %5
  ret i64 %10
}

julia> @code_llvm debuginfo=:none g(UInt(1), 1)

define i64 @julia_g_15989(i64, i64) {
top:
  %2 = and i64 %1, 63
  %3 = shl i64 %0, %2
  ret i64 %3
}
```

Comment (chethega) [reply base/broadcast.jl]: As @maxbennedich recently pointed out, `f(a,b)=a<<b` and `g(a,b)= a<<(b&63)` compile to quite different code for integers on x86. If the second argument is larger or equal than the bitwidth, then julia wants to make the value zero, while the native instruction only considers the relevant bits (implements `g`). Hence, the `g` variant is faster on x86 if the compiler fails to prove that the second argument is between 0 and 63. 

What I really want here is`shl(a::UInt64,b::Int64)=Core.Intrinsics.llvmcall("%res =shl i64 %0, %1\n ret i64 %res", UInt64, Tuple{UInt64, Int64}, a, b)`, i.e. I want to express that shifting left by more than 63 bit is undefined behavior and llvm can emit whatever instruction does the job fastest on the target arch. I don't know what kind of idiom wins for arm or power.

You are definitely right that this is very smelly code.

File: base/broadcast.jl [hunk 2]
```diff
@@ -974,6 +974,32 @@ preprocess(dest, x) = extrude(broadcast_unalias(dest, x))
...
 
+# Performance optimization: for BitVector outputs, we cache the result
+# in a 64-bit register before writing into memory (to bypass LSQ)
+@inline function copyto!(dest::BitVector, bc::Broadcasted{Nothing})
+    axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
+    ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
+    destc = dest.chunks
+    bcp = preprocess(dest, bc)
+    length(bcp) <= 0 && return dest
+    len = Base.num_bit_chunks(length(bcp))
```
Comment (vtjnash) [inline base/broadcast.jl:985]:
```suggestion
    len = Base.num_bit_chunks(Int(length(bcp)))
```

**Pull request merged**